Count stashed errors again

nnethercote opened the pull request:
Stashed diagnostics are such a pain. Their "might be emitted, might not" semantics messes with lots of things.

#120828 and #121206 made some big changes to how they work, improving some things, but still leaving some problems, as seen by the issues caused by #121206. This PR aims to fix all of them by restricting them in a way that eliminates the "might be emitted, might not" semantics while still allowing 98% of their benefit. Details in the individual commit logs.

r? @oli-obk

nnethercote committed:
Overhaul how stashed diagnostics work, again.

Stashed errors used to be counted as errors, but could then be
cancelled, leading to `ErrorGuaranteed` soundness holes. #120828 changed
that, closing the soundness hole. But it introduced other difficulties
because you sometimes have to account for pending stashed errors when
making decisions about whether errors have occured/will occur and it's
easy to overlook these.

This commit aims for a middle ground.
- Stashed errors (not warnings) are counted immediately as emitted
  errors, avoiding the possibility of forgetting to consider them.
- The ability to cancel (or downgrade) stashed errors is eliminated, by
  disallowing the use of `steal_diagnostic` with errors, and introducing
  the more restrictive methods `try_steal_{modify,replace}_and_emit_err`
  that can be used instead.

Other things:
- `DiagnosticBuilder::stash` and `DiagCtxt::stash_diagnostic` now both
  return `Option<ErrorGuaranteed>`, which enables the removal of two
  `delayed_bug` calls and one `Ty::new_error_with_message` call. This is
  possible because we store error guarantees in
  `DiagCtxt::stashed_diagnostics`.
- Storing the guarantees also saves us having to maintain a counter.
- Calls to the `stashed_err_count` method are no longer necessary
  alongside calls to `has_errors`, which is a nice simplification, and
  eliminates two more `span_delayed_bug` calls and one FIXME comment.
- Tests are added for three of the four fixed PRs mentioned below.
- `issue-121108.rs`'s output improved slightly, omitting a non-useful
  error message.

Fixes #121451.
Fixes #121477.
Fixes #121504.
Fixes #121508.

nnethercote committed:
Don't cancel stashed `OpaqueHiddenTypeMismatch` errors.

This gives one extra error message on one test, but is necessary to fix
bigger problems caused by the cancellation of stashed errors.

(Note: why not just avoid stashing altogether? Because that resulted in
additional output changes.)

nnethercote committed:
Don't cancel stashed `TraitMissingMethod` errors.

This gives one extra error message on two tests, but is necessary to fix
bigger problems caused by the cancellation of stashed errors.

(Note: why not just avoid stashing altogether? Because that resulted in
additional output changes.)

nnethercote committed:
Mark some once-again-unreachable paths as unreachable.

This undoes the changes from #121482 and #121586, now that stashed errors
will trigger more early aborts.

nnethercote committed:
Stop miri if delayed bugs are present.

Seems wise, since it shouldn't proceed in that case.

oli-obk commented on compiler/rustc_middle/src/ty/mod.rs:
```diff
@@ -846,12 +846,17 @@ impl<'tcx> OpaqueHiddenType<'tcx> {
+            // cancelling stashed diagnostics is no longer allowed because it
+            // causes problems when tracking whether errors have actually
+            // occurred.
+            diag.emit();
```
would we get the same benefits by turning it into a delayed bug?

estebank replied:
+1 on this.

nnethercote replied:
Unfortunately, the subequent commits forbid that. The reason being that it will mess with the error counts. A stashed error gets counted by `DiagCtxt::err_count` and `DiagCtxt::err_count_excluding_lint_erss`. If it could later be downgraded to a delayed bug then that would no longer be counted by those methods, and the error count would go down. That's not as bad as conjuring an `ErrorGuaranteed` out of nothing, but it's still the kind of laxness that I am desperately trying to eliminate from the stashed diagnostics implementation.

In general, the old stashed diagnostics implementation was wildly unsafe and overpowered, and I'm having to introduce some limitations to fix the problems.

Alexendoo commented on src/tools/clippy/tests/ui/unreachable-array-or-slice.rs:
Ah I missed https://github.com/rust-lang/rust/issues/121477, would be good to put it in `src/tools/clippy/tests/ui/crashes` in that case

nnethercote replied:
:+1:

estebank commented on compiler/rustc_privacy/src/lib.rs:
```diff
@@ -988,10 +988,7 @@ impl<'tcx> Visitor<'tcx> for NamePrivacyVisitor<'tcx> {
-                self.tcx.dcx().span_delayed_bug(expr.span, "no adt_def for expression");
-                return;
-            };
+            let adt = self.typeck_results().expr_ty(expr).ty_adt_def().unwrap();
```
Slightly concerned that this will lead to an ICE that used not to happen (`span_delayed_bug` has been the go-to solution for "this is invalid state but other errors will likely have been emitted so an ICE isn't necessary).

nnethercote replied:
It's just reverting to what was there before #121482.

estebank commented on compiler/rustc_middle/src/ty/mod.rs:
```diff
@@ -846,13 +846,15 @@ impl<'tcx> OpaqueHiddenType<'tcx> {
-        {
-            diag.cancel();
-        }
+        // We used to cancel here for slightly better error messages, but
+        // cancelling stashed diagnostics is no longer allowed because it
+        // causes problems when tracking whether errors have actually
+        // occurred.
+        tcx.sess.dcx().try_steal_modify_and_emit_err(
+            tcx.def_span(opaque_def_id),
+            StashKey::OpaqueHiddenTypeMismatch,
+            |_err| {},
+        );
```
This change is fine.

estebank commented on compiler/rustc_lint/src/array_into_iter.rs:
```diff
@@ -70,15 +70,11 @@ impl<'tcx> LateLintPass<'tcx> for ArrayIntoIter {
-            {
-                return;
-            }
+            let def_id = cx.typeck_results().type_dependent_def_id(expr.hir_id).unwrap();
```
I'm very weary of `.unwrap()` in the compiler. I've had cases where I leaned on it and later changes caused my expectations to not be met.

nnethercote replied:
And this is just reverting to what was there before #121586.

I understand the purpose of `span_delayed_bug`, but it shouldn't be used on genuinely unreachable code paths, right? I don't know if you saw #121208 where I deliberately changed a bunch of definitely/probably/possibly unreachable `span_delayed_bug` calls to `span_bug`... some of them were found with subsequent fuzzing to be reachable, which means we got test cases for those paths.

estebank reviewed (approved):
The code and output changes look reasonable to me. Left some nitpicks, but r=me on the approach.

nnethercote commented:
I have updated to move the clippy test. I haven't made any other changes, not sure if my responses to the other comments will satisfy the reviewers.

(And apologies for messing up the syntax on the review request in the original description, it was meant to be for @oli-obk, but luck would have it that @estebank is a good reviewer for this PR as well.)

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8071576845/job/22051489535) [(plain)](https://github.com/rust-lang/rust/commit/21ce4d1caa81ded6d1aa497ceb7721850e45da10/checks/22051489535/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_e4944055-d6fb-44a9-b0e5-4f8efb620de3
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=count-stashed-errs-again
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_e4944055-d6fb-44a9-b0e5-4f8efb620de3
GITHUB_REF=refs/pull/121669/merge
GITHUB_REF_NAME=121669/merge
GITHUB_REF_PROTECTED=false
---
Location:
    /cargo/registry/src/index.crates.io-6f17d22bba15001f/ui_test-0.22.2/src/lib.rs:395:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

FAILED TEST: tests/ui/crashes/unreachable-array-or-slice.rs
command: CLIPPY_CONF_DIR="tests" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-ea5fb8c487f68a5a.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-e4b3ef73bbd8fea6.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-4a90d522ee4fbe1d.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-c9f5b36cc1e7e065.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-ae846d6788d6a9f1.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-9152caa60d22e5d6.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-389517e109d775ca.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-30832ea3988a6e92.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-39bd2cd32ebc6796.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-f2adc3bd2db35390.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps/libserde_derive-81cbf1c0cde13f71.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-b4a0c687f454934f.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-eddc5a1e301140d8.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui/crashes" "tests/ui/crashes/unreachable-array-or-slice.rs" "--edition" "2021"

error: actual output differed from expected
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/crashes/unreachable-array-or-slice.stderr` to the actual output
--- tests/ui/crashes/unreachable-array-or-slice.stderr
 error[E0433]: failed to resolve: `Self` is only available in impls, traits, and type definitions
-  --> $DIR/unreachable-array-or-slice.rs:4:9
+  --> tests/ui/crashes/unreachable-array-or-slice.rs:4:9
    |
    |
 LL |     let Self::anything_here_kills_it(a, b, ..) = Foo(5, 5, 5, 5);
 
 For more information about this error, try `rustc --explain E0433`.



full stderr:
error[E0433]: failed to resolve: `Self` is only available in impls, traits, and type definitions
##[error]  --> tests/ui/crashes/unreachable-array-or-slice.rs:4:9
   |
LL |     let Self::anything_here_kills_it(a, b, ..) = Foo(5, 5, 5, 5);
   |         ^^^^ `Self` is only available in impls, traits, and type definitions
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0433`.

---

Build completed unsuccessfully in 0:01:47


command did not execute successfully: cd "/checkout" && env -u MAKEFLAGS -u MFLAGS AR_x86_64_unknown_linux_gnu="ar" CARGO_INCREMENTAL="0" CARGO_PROFILE_RELEASE_DEBUG="0" CARGO_PROFILE_RELEASE_DEBUG_ASSERTIONS="true" CARGO_PROFILE_RELEASE_OVERFLOW_CHECKS="true" CARGO_PROFILE_RELEASE_STRIP="false" CARGO_TARGET_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools" CC_x86_64_unknown_linux_gnu="sccache cc" CFG_COMPILER_BUILD_TRIPLE="x86_64-unknown-linux-gnu" CFG_COMPILER_HOST_TRIPLE="x86_64-unknown-linux-gnu" CFG_RELEASE="1.78.0-nightly" CFG_RELEASE_CHANNEL="nightly" CFG_RELEASE_NUM="1.78.0" CFG_VERSION="1.78.0-nightly (2bba296b1 2024-02-27)" CFG_VER_DATE="2024-02-27" CFG_VER_HASH="2bba296b1d6de8abe2d152577d5d6018f8cf22e0" CFLAGS_x86_64_unknown_linux_gnu="-ffunction-sections -fdata-sections -fPIC -m64" CXXFLAGS_x86_64_unknown_linux_gnu="-ffunction-sections -fdata-sections -fPIC -m64" CXX_x86_64_unknown_linux_gnu="sccache c++" DOC_RUST_LANG_ORG_CHANNEL="https://doc.rust-lang.org/nightly" HOST_LIBS="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release" LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" LIBC_CHECK_CFG="1" LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm/lib" LIBZ_SYS_STATIC="1" LZMA_API_STATIC="1" RANLIB_x86_64_unknown_linux_gnu="ar s" REAL_LIBRARY_PATH_VAR="LD_LIBRARY_PATH" RUSTBUILD_NATIVE_DIR="/checkout/obj/build/x86_64-unknown-linux-gnu/native" RUSTC="/checkout/obj/build/bootstrap/debug/rustc" RUSTC_BOOTSTRAP="1" RUSTC_BREAK_ON_ICE="1" RUSTC_ERROR_METADATA_DST="/checkout/obj/build/tmp/extended-error-metadata" RUSTC_HOST_FLAGS="-Zunstable-options --check-cfg=cfg(bootstrap)" RUSTC_INSTALL_BINDIR="bin" RUSTC_LIBDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTC_LIB_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTC_LINT_FLAGS="-Wrust_2018_idioms -Wunused_lifetimes -Dwarnings" RUSTC_REAL="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_SNAPSHOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_SNAPSHOT_LIBDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTC_STAGE="2" RUSTC_SYSROOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" RUSTC_TEST_SUITE="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUSTC_TLS_MODEL_INITIAL_EXEC="1" RUSTC_VERBOSE="0" RUSTC_WRAPPER="/checkout/obj/build/bootstrap/debug/rustc" RUSTDOC="/checkout/obj/build/bootstrap/debug/rustdoc" RUSTDOCFLAGS="--cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options --check-cfg=cfg(bootstrap) --check-cfg=cfg(parallel_compiler) --check-cfg=cfg(rust_analyzer) --check-cfg=cfg(no_btreemap_remove_entry) --check-cfg=cfg(crossbeam_loom) --check-cfg=cfg(span_locations) --check-cfg=cfg(rustix_use_libc) --check-cfg=cfg(emulate_second_only_system) --check-cfg=cfg(windows_raw_dylib) -Dwarnings -Wrustdoc::invalid_codeblock_attributes --crate-version 1.78.0-nightly\t(2bba296b1\t2024-02-27)" RUSTDOC_LIBDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTDOC_REAL="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" RUSTFLAGS="--cfg=windows_raw_dylib -Csymbol-mangling-version=v0 -Zunstable-options --check-cfg=cfg(bootstrap) --check-cfg=cfg(parallel_compiler) --check-cfg=cfg(rust_analyzer) --check-cfg=cfg(no_btreemap_remove_entry) --check-cfg=cfg(crossbeam_loom) --check-cfg=cfg(span_locations) --check-cfg=cfg(rustix_use_libc) --check-cfg=cfg(emulate_second_only_system) --check-cfg=cfg(windows_raw_dylib) -Zmacro-backtrace -Csplit-debuginfo=off -Clink-args=-Wl,-z,origin -Clink-args=-Wl,-rpath,$ORIGIN/../lib -Zunstable-options" RUST_TEST_THREADS="16" SYSROOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" TERM="xterm" __CARGO_DEFAULT_LIB_METADATA="nightlytool-rustc" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "--release" "-Zcheck-cfg" "-Zbinary-dep-depinfo" "-j" "16" "--locked" "--color" "always" "--manifest-path" "/checkout/src/tools/clippy/Cargo.toml" "--" "--quiet"

stdout ----

stderr ----
```

</details>

nnethercote committed:
Reinstate `emit_stashed_diagnostics` in `DiagCtxtInner::drop`.

I removed it in #121206 because I thought thought it wasn't necessary.
But then I had to add an `emit_stashed_diagnostics` call elsewhere in
rustfmt to avoid the assertion failure (which took two attempts to get
right, #121487 and #121615), and now there's an assertion failure in
clippy as well (https://github.com/rust-lang/rust-clippy/issues/12364).

So this commit just reinstates the call in `DiagCtxtInner::drop`. It
also reverts the rustfmt changes from #121487 and #121615, though it
keeps the tests added for those PRs.

nnethercote commented:
I added one more commit, reinstating the `emit_stashed_diagnostics` call in `DiagCtxtInner::drop`, to fix all the assertion failures that we were getting (details in the commit message). That should fix all the problems that #121206 caused, I hope.

oli-obk commented:
@bors r=estebank

fmease commented:
@bors r=estebank

bors commented:
:pushpin: Commit 1c5d1c7d689a6dec049c30a1e1aa8011bb4a7352 has been approved by `estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank 1c5d1c7d689a6dec049c30a1e1aa8011bb4a7352 -->
<!-- homu: {"type":"Approved","sha":"1c5d1c7d689a6dec049c30a1e1aa8011bb4a7352","approver":"estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #121489) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
I rebased.

@bors r=estebank

bors commented:
:pushpin: Commit 82961c0abcd7b9ea73fb87fd049e2e853abd5787 has been approved by `estebank`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=estebank 82961c0abcd7b9ea73fb87fd049e2e853abd5787 -->
<!-- homu: {"type":"Approved","sha":"82961c0abcd7b9ea73fb87fd049e2e853abd5787","approver":"estebank","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
